includes/subscriptions/class-wc-payments-product-service.php

```diff
@@ -115,40 +131,78 @@ public static function get_wcpay_product_hash( WC_Product $product ): string {
+	public function get_or_create_wcpay_product_id( WC_Product $product, $test_mode = null ): string {
 		// If the subscription product doesn't have a WC Pay product ID, create one.
-		if ( ! self::has_wcpay_product_id( $product, $test_mode ) ) {
+		if ( ! $this->has_wcpay_product_id( $product, $test_mode ) ) {
 			$is_current_environment = null === $test_mode || WC_Payments::mode()->is_test() === $test_mode;
 
 			// Only create a new wcpay product if we're trying to fetch a wcpay product ID in the current environment.
 			if ( $is_current_environment ) {
-				WC_Payments_Subscriptions::get_product_service()->create_product( $product );
+				$this->create_product( $product );
 			}
 		}
 
-		return $product->get_meta( self::get_wcpay_product_id_option( $test_mode ), true );
+		return $product->get_meta( self::get_wcpay_product_id_meta_key( $test_mode ), true );
 	}
 
 	/**
-	 * Gets the WC Pay product ID associated with a WC product.
+	 * Get the WCPay product ID for an item type.
 	 *
-	 * @param string $type The item type to create a product for.
-	 * @return string       The item's WCPay product id.
+	 * @param string $type The item type.
+	 *
+	 * @return string The WCPay product ID.
+	 * @throws API_Exception
 	 */
 	public function get_wcpay_product_id_for_item( string $type ): string {
-		$sanitized_type  = self::sanitize_option_key( $type );
-		$option_key_name = self::get_wcpay_product_id_option() . '_' . $sanitized_type;
-		if ( ! get_option( $option_key_name ) ) {
-			$this->create_product_for_item_type( $sanitized_type );
+		$sanitized_type   = self::sanitize_option_key( $type );
+		$option_key_name  = self::get_wcpay_product_id_option() . '_' . $sanitized_type;
+		$wcpay_product_id = get_option( $option_key_name );
+
+		// Case 1: No product found, create a new one.
+		if ( ! $wcpay_product_id ) {
+			return $this->create_product_for_item_type( $sanitized_type );
 		}
-		return get_option( $option_key_name );
+
+		// For existing products, check the linked account.
+		$linked_option_key = $option_key_name . '_linked_to';
+		$linked_account_id = get_option( $linked_option_key );
+		$stripe_account_id = $this->get_stripe_account_id();
+
+		// Case 2: Product exists but linked account doesn't, validate and update if needed.
+		if ( ! $linked_account_id ) {
+			try {
+				// Validate that the product exists for the current account.
+				$existing_product = $this->payments_api_client->get_product_by_id( $wcpay_product_id );
+
+				if ( $existing_product ) {
+					// Product exists, save with current account ID.
+					$this->save_wcpay_product_data( $wcpay_product_id, $stripe_account_id, $sanitized_type );
+					return $wcpay_product_id;
+				} else {
+					// Product doesn't exist, create new one.
+					return $this->create_product_for_item_type( $sanitized_type );
+				}
+			} catch ( \Exception $e ) {
+				// Error occurred, create new product.
+				return $this->create_product_for_item_type( $sanitized_type );
+			}
```

Comment:
minor: I wonder if the merchant should be aware of the fact that we needed to create a new product. I guess not, but maybe we could write something to the logs about this action? Perhaps using the debug level, so it can be easier to troubleshoot if needed?